crates/chia-bls/src/cached_bls.rs

```diff
@@ -0,0 +1,295 @@
+// This cache is a bit weird because it's trying to account for validating
+// mempool signatures versus block signatures. When validating block signatures,
+// there's not much point in caching the pairings because we're probably not going
+// to see them again unless there's a reorg. However, a spend in the mempool
+// is likely to reappear in a block later, so we can save having to do the pairing
+// again. So caching is primarily useful after "catch-up" (fast sync?) is done and
+// we're monitoring the mempool in real-time.
+
+use crate::aggregate_verify as agg_ver;
+use crate::gtelement::GTElement;
+use crate::hash_to_g2;
+use crate::PublicKey;
+use crate::Signature;
+use lru::LruCache;
+use sha2::{Digest, Sha256};
+use std::borrow::Borrow;
+use std::collections::HashMap;
+use std::num::NonZeroUsize;
+
+#[cfg(feature = "py-bindings")]
+use pyo3::types::{PyBool, PyInt, PyList};
+#[cfg(feature = "py-bindings")]
+use pyo3::{pyclass, pymethods, PyResult};
+
+#[cfg_attr(feature = "py-bindings", pyclass(name = "BLSCache"))]
+pub struct BLSCache {
+    cache: LruCache<[u8; 32], GTElement>,
+}
+
+impl Default for BLSCache {
+    fn default() -> Self {
+        Self::new(50000)
+    }
+}
+
+impl BLSCache {
+    pub fn new(cache_size: usize) -> BLSCache {
+        let cache: LruCache<[u8; 32], GTElement> =
+            LruCache::new(NonZeroUsize::new(cache_size).unwrap());
+        Self { cache }
+    }
+
+    pub fn generator(cache_size: Option<usize>) -> Self {
+        let cache: LruCache<[u8; 32], GTElement> =
+            LruCache::new(NonZeroUsize::new(cache_size.unwrap_or(50000)).unwrap());
+        Self { cache }
+    }
+
+    pub fn get_pairings<P: Borrow<[[u8; 48]]>, M: AsRef<[u8]>>(
+        &mut self,
+        pks: &P,
+        msgs: &[M],
+        force_cache: bool,
+    ) -> Vec<GTElement> {
+        let mut pairings: Vec<Option<GTElement>> = vec![];
```

Comment:
this temporary vector, as well as the `ret` returned vector are probably pretty expensive since they allocate memory and increase the memory footprint of this function. Since `get_pairings()` is only used internally, we don't necessarily need to maintain this interface. For example, we could instead pass in a mutable `blst_pairing` object (or perhaps some abstraction of it) and directly apply the cached `GTElements` with `blst_pairing_raw_aggregate()`. That would very likely be a lot more efficient.